docs/Community/Governance.md

```diff
@@ -0,0 +1,137 @@
+or the CGC.
+
+If there is conflict, either within or between SIGs, or between any members of
+the community, wherein it is desirable, the issue can be raised to the CGC.
+
+Technical Discussions
+---------------------
+
+Technical discussions generally happen in one of these places:
+
+* [on the GitHub project](https://github.com/openenclave/openenclave), either in
+  an Issue or a Pull Request
+* in the public meetings of Committees, Special Interest Groups, or Working Groups, which can be found via the [community calendar](https://lists.confidentialcomputing.io/calendar).
+* on the community mailing list or chat server, which is detailed in the [Comminucations](Communication.md) document
+
+While we strive for openness and transparency, it is natural that some
+discussions will also happen in private channels, or in impromptu face-to-face
+conversations (for those who work in the same building, or bump into each other
+at a conference). The outcome of such discussions should be considered
+non-binding until the information is shared through an open medium and relevant
+stake-holders have opportunity to contribute to the discussion.
+
+All SIG and WG leaders are responsible for creating and publishing minutes from
+the meetings they hold, ensuring that the meetings are open and inclusive, and
+that decisions they make are fair and transparent.
+
+If a private discussion pertains to an open issue or PR, and results in a
+decision, that item should be updated and reference given to the conversation.
+
+Issue/PR Review Process
+-----------------------
+
+The community strives to triage and respond to all incoming issues within one
+week. If you haven't heard anything by then, or an issue or PR is not getting
+any response within a reasonable time after it is triaged, please feel free to
+remind us with a ping on the thread.
+
+Security issues should be reported through a separate channel, and
+will receive a response within 24 hours. See [Reporting Security
+Issues](Contributing.md#reporting-security-issues).
+
+Committers and Contributors
+---------------------------
+
+At present, we maintain a list of [Project Committers](Committers.md) along with
+descriptions of their knowledge areas.
+
+This will be updated and replaced with a more scaleable approach to
+commit-rights management.
+
+Accepting Contributions
+-----------------------
+
+SIGs will generally accept changes that improve the project and that align with the
+[Open Enclave roadmap](https://github.com/openenclave/openenclave/projects).
+
+All contributions are expected to be reviewed by at least one SIG lead other
+than the contributor, and have no significant objections from other SIG leads
+within the same SIG.
+
+If a change requires the approval of multiple SIGs (e.g., because it affects
+code in both SIG domains), then approval from both SIGs is required.
+
+If a change only requires one approval from one SIG, and leads of another SIG
```

Comment:
The Committers.md still has old text that points here, but still says:
> However, it is up to each committer to determine who should review which PR, and
> when to merge it. It is up to each committer to decide [when to accept a contribution](Governance.md#Accepting-Contributions)

A couple of clarifications:
* Is the implication that every committer is a member of a SIG? Is a distinction being made between a contributor as a SIG member vs. a SIG lead?
* If not, what would be the relationship between the approval of individual committer and a SIG member? 
   * e.g. in our current 2 approvals required policy, it could be any 1 committer plus a SIG lead, but not 2 non-SIG committers?

Reply:
This is an area that I would like to discuss and clarify more. Agree that it's not clear right now, even with the draft changes I propose.

Here's my thinking:
- remove the idea of "committers" that is present today, replacing that with "approvers" and "reviewers"
- using automation (eg, prow & lgtm, as mentioned above), distribute fine-grained rights management through OWNERS files in each directory
- "reviewers" are able to leave "lgtm" comments, which the bots aggregate to manage github tags and gating tests
- "approvers" are able to leave "approve" comments, which, when there are required # of approvals from the right people, then trigger a merge to be performed by the bots
- reduce # of people who actually have write access (eg, can "force push") to the repo to the absolute bare minimum of 3 - 4 people. Less than that risks project continuity if there's an accident

See https://prow.k8s.io/command-help for listing of some of the commands k8s uses. Commands are fully customizable as well.

Following this approach, a commit would need to have two "lgtm" and one "approve" comments, from people who have the correct permissions to review and approve changes to that portion of the code base. This is only a proposal -- needing three sets of eyes on a change may be too many for the current team size. My main point is that the tooling would help us here by introducing an explicit and transparent review structure, delegating both authority and autonomy to domain-owners, and thereby allowing the review workload to spread out more.

Reply:
I like the structure of that, thanks for elaborating on the thinking there! I'm primarily looking for clarity on what changes in processes between what exists today and how to get people to transition to different practices described in the new text; a couple of additional clarifications and thoughts:

* I mentally map "approvers" to "committers" today in the sense that these are the individuals identified in the CODEOWNERS file today that would map to OWNERS in prow. 
    * Is "reviewers" a special category in that case? Prow describes them as any collaborators on the project, and so in terms of access rights, these would just be equivalent to general contributors in the OE project? i.e. there is no special process to identify or approve individuals as being "reviewers".
    * My impression (please correct) is the difference between "approvers" and "committers" is "approvers" do not have the same access privileges as implied by "committers" today:
        * "approvers" do not have elevated project privileges like write or force push. Their authority to merge is controlled through automation (i.e. prow)
        * Do "approvers" get heightened privileges in configuring prow as well? (i.e. committers today through their privileges on the project also have elevated privileges on bors)
        * If the idea is to restrict project privileges to a handful of individuals, we should probably also describe the process for selecting them since it sounds like this is likely a subset of the CGC members? That could probably be deferred to a future CGC meeting discussion. 
* Yeah, 3 individuals per PR is probably more than the current team size can keep up with right now, given the code velocity to date. I would suggest we keep the current requirement of two approvals or an LGTM and an approval until the project scales up. This is likely even more true as we fragment review responsibilities to individual SIGs.

Reply:
> "approvers" do not have elevated project privileges like write or force push. Their authority to merge is controlled through automation (i.e. prow)

Correct

> Do "approvers" get heightened privileges in configuring prow as well? (i.e. committers today through their privileges on the project also have elevated privileges on bors)

The list of approvers, being determined by an OWNERS file in a given directory, becomes self-governing through Prow. Those approvers (and anyone who supercedes them) become the folks who can change the list of approvers within that OWNERs file.

If prow configuration itself were managed through code in a directory, then the approvers listed in the OWNERS file in _that_ directory would be the list of people who can control prow's configuration.

In this way, many people may be able to approve project changes _without_ also being able to control the prow configuration. The infra team would control the infrastructure. 

> If the idea is to restrict project privileges to a handful of individuals, we should probably also describe the process for selecting them since it sounds like this is likely a subset of the CGC members? That could probably be deferred to a future CGC meeting discussion.

Agreed. Worthwhile discussion for a future CGC meeting, and/or a thread on the CGC ML.

As the project grows, gains more contributors and more attention, keeping the set of people who retain project privileges (eg, can force push) to a very small set increases in importance. This is the set of people who are trusted with the power to break the project history (and who, clearly, would not do that unless absolutely necessary).

I'm not concerned with anyone on the current list; rather, my concern is that, with the process that is in place today, I don't see a way to expand the approver list without expanding the committer list, or to delegate domain-specific approval rights.